File: Simplex.kotlin/src/main/kotlin/kr/co/tmon/simplex/extensions/ActionExtension.kt
package kr.co.tmon.simplex.extensions

import kr.co.tmon.simplex.actions.*
import io.reactivex.Observable

internal fun <TParam, TResult> IAction<TResult>.toObservable(param: TParam): Observable<TResult> {
    @Suppress("UNCHECKED_CAST")
    val parameterizedAction =  this as? IParameterizedAction<TParam, TResult>
    return parameterizedAction?.process(param) ?: this.process()
}

File: Simplex.Core/Extensions/ActionExtension.cs
using System;
using System.Collections.Generic;
using System.Text;
using Tmon.Simplex.Actions;

namespace Tmon.Simplex.Extensions
{
    public static class ActionExtension
    {
        internal static IObservable<TResult> InvokeProcess<TParam, TResult>(this IAction<TResult> source, TParam param)
        {
            Func<TParam, IObservable<TResult>> process;
            if (source is IAction<TParam, TResult> paramAction)
                process = (p => paramAction.Process(p));
            else
                process = (p => source.Process());

            return process.Invoke(param);
        }
    }
}
